cmd/zfs/zfs_main.c

```diff
@@ -7289,6 +7389,230 @@ zfs_do_change_key(int argc, char **argv)
+ *    -p: Specify the expected project ID for checking tree quota. It needs
+ *        to be used together with "-r".
+ *    -0: Print entries found by -c with a trailing NUL instead of a newline,
+ *        like "find -print0". It needs to be specified together with "-r".
+ *
+ * 3) zfs project -C [-r] [-k] <file|directory ...>
+ *    Clear project ID inherit flag for the given file or directory.
+ *    -r: All descendants' (of the directory) inherit flag will be cleared.
+ *    -k: Keep the porject ID unchanged. If not specified, the directory's
+ *        (and all its descendants' for "-r" specified) project ID will be
+ *        reset as zero.
+ *
+ * 4) zfs project -s [-r] [-p id] <file|directory ...>
+ *    Set project ID inherit flag for the given file or directory.
+ *    -r: Set project ID with the directory's project ID for all its
+ *        descendants. For descendant directories, also set inherit flag.
+ *    -p: The file or directory's (and all its descendants' if "-r" option
+ *        is specified) project ID will be set with the given value.
+ */
+static int
+zfs_do_project(int argc, char **argv)
+{
+	boolean_t keep = B_FALSE;
+	boolean_t newline = B_TRUE;
+	boolean_t dironly = B_FALSE;
+	boolean_t recursive = B_FALSE;
+	uint64_t projid = ZFS_INVALID_PROJID;
+	int ret = 0, c, op = 0;
+
+	if (argc < 3)
+		usage(B_FALSE);
+
+	while ((c = getopt(argc, argv, "p:cCsdkr0")) != -1) {
+		switch (c) {
+		case 'c':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_CHECK;
+			break;
+		case 'C':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_CLEAR;
+			break;
+		case 's':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_SET;
+			break;
+		case 'd':
+			if (recursive) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-d' and '-r' together\n"));
+				usage(B_FALSE);
+			}
+
+			dironly = B_TRUE;
+			break;
+		case 'k':
+			keep = B_TRUE;
+			break;
+		case 'r':
+			if (dironly) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-d' and '-r' together\n"));
+				usage(B_FALSE);
+			}
+
+			recursive = B_TRUE;
+			break;
+		case 'p':
+			projid = strtoull(optarg, NULL, 0);
+			if (projid == ZFS_INVALID_PROJID) {
+				(void) fprintf(stderr,
+				    gettext("invalid project ID\n"));
+				usage(B_FALSE);
+			}
+
+			break;
+		case '0':
+			newline = B_FALSE;
+			break;
+		default:
+			(void) fprintf(stderr, gettext("invalid option '%c'\n"),
+			    optopt);
+			usage(B_FALSE);
+		}
+	}
+
+	switch (op) {
+	case ZFS_PROJECT_CHECK:
+		if (!recursive) {
+			if (projid != ZFS_INVALID_PROJID) {
```

Comment:
This has following problem:

zfs project -c -p 1 <regular_file>

This means we could not check a regualr file without -r, but with -r,
it will return "'-r' option on non-dir target"

Reply:
Also please remove "'-r' option on non-dir target", i think it should be fine, to use -r to operate on 
a regular file, just return and works fine is ok.

Reply:
That is one possible solution. But it will make the "-r" semantics more confused. Let's consider whether it is possible to add another sub-commands set "zfs project -l [-d] <file|directory ...>"

Reply:
"-c -p1" is used for checking whether the target tree has the specified project ID "1" and inherit flag. It only works on directory, not regular file. If you want to show regular-file's project ID, please use "zfs project -c" directly without other options.